taichi/analysis/gather_uniquely_accessed_pointers.cpp

```diff
@@ -0,0 +1,157 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/visitors.h"
+#include <algorithm>
+
+TLANG_NAMESPACE_BEGIN
+
+class LoopUniqueStmtSearcher : public BasicStmtVisitor {
+ private:
+  std::unordered_set<Stmt *> loop_invariant_;
```

Comment:
super nit: `loop_constants_`? (From what I understand, this does a simple constant propagation?)